File: Diagrammatize.Design.Pattern/code/python/Decorator/A1/FullBorder.py
from Border import Border

class FullBorder(Border):
    def __init__(self, display):
        super().__init__(display)

    def getColumns(self):
        return 1 + self.display.getColumns() + 1

    def getRows(self):
        return 1 + self.display.getRows() + 1

    def getRowText(self, row):
        if row == 0:
            return "+" + self.makeLine('-', self.display.getColumns()) + "+"
        elif row == self.display.getRows() + 1:
            return "+" + self.makeLine('-', self.display.getColumns()) + "+"
        else: 
            return "|" + self.display.getRowText(row - 1) + "|"

    def makeLine(self, ch, count):
        return str(ch)*count

File: Diagrammatize.Design.Pattern/code/cxx/Decorator/Sample/FullBorder.cpp
#include "FullBorder.hpp"
#include <sstream>

FullBorder::FullBorder(Display* display): Border(display) {
}

int FullBorder::getColumns() {
    return 1 + display->getColumns() + 1;
}

int FullBorder::getRows() {
    return 1 + display->getRows() + 1;
}

std::string FullBorder::getRowText(int row) {
    if (row == 0) {                                                 // 上边框
        return "+" + makeLine('-', display->getColumns()) + "+";
    } else if (row == display->getRows() + 1) {                      // 下边框
        return "+" + makeLine('-', display->getColumns()) + "+";
    } else {                                                        // 其他边框
        return "|" + display->getRowText(row - 1) + "|";
    }
}

std::string FullBorder::makeLine(char ch, int count) {
    return std::string(count, ch);
}
